bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/HandlerProxy.java

```diff
@@ -475,4 +481,20 @@ public String getAttributeName() {
 	public IHandler getHandler() {
 		return handler;
 	}
+
+	@Override
+	public void addState(String stateId, State state) {
+		super.addState(stateId, state);
```

Comment:
Now that we handle state addition correctly, `handleStateChange()` should be done internally by handlers supporting `IObjectWithState` and `IStateListener` protocols. However, as this functionality was broken for ten years, it is unlikely, that clients properly support the protocol. For this reason clients that support the protocol will receive two events - "state added" and "state changed" instead of the original design, where "state added" was supposed to be enough.

Reply:
Removed `IStateListener` branch, as it could never be called due to bugs in `E4HandlerProxy`.